File: hackerrank/problem solving/algorithms/1 - warmup/3 - compare the triplets/compare_the_triplets.go
// https://www.hackerrank.com/challenges/compare-the-triplets/problem?isFullScreen=true

package main

import "fmt"

func main() {
    const N = 3
    tripletA := readNums(N)
    tripletB := readNums(N)
    scoreA, scoreB := compareTriplets(tripletA, tripletB)
    fmt.Println(scoreA, scoreB)
}

func readNums(n int) []int {
    arr := make([]int, n)
    for i := range arr {
        _, _ = fmt.Scan(&arr[i])
    }
    return arr
}

// n: length of a and b = 3
// T: O(3) => O(1)
// S: O(1) extra space
func compareTriplets(tripletA, tripletB []int) (int, int) {
    scoreA, scoreB := 0, 0

    for i := range tripletA {
        if tripletA[i] > tripletB[i] {
            scoreA++
        } else if tripletB[i] > tripletA[i] {
            scoreB++
        }
    }

    return scoreA, scoreB
}

File: hackerrank/problem solving/algorithms/1 - warmup/3 - compare the triplets/compare_the_triplets.rs
// https://www.hackerrank.com/challenges/compare-the-triplets/problem?isFullScreen=true

fn main() {
    const N: usize = 3;
    let triplet_a: Vec<i32> = read_nums(N);
    let triplet_b: Vec<i32> = read_nums(N);
    let (score_a, score_b) = compute_triplet_scores(triplet_a, triplet_b);
    println!("{} {}", score_a, score_b);
}

fn read_nums(n: usize) -> Vec<i32> {
    (0..n).map(|_| text_io::read!()).collect()
}

// n: length of a and b = 3
// T: O(3) => O(1)
// S: O(1) extra space
fn compute_triplet_scores(triplet_a: Vec<i32>, triplet_b: Vec<i32>) -> (i32, i32) {
    let (mut score_a, mut score_b) = (0, 0);

    for (ta, tb) in triplet_a.iter().zip(triplet_b.iter()) {
        if ta > tb {
            score_a += 1;
        } else if tb > ta {
            score_b += 1;
        }
    }

    (score_a, score_b)
}
